corehq/motech/value_source.py

```diff
@@ -451,6 +449,124 @@ class CasePropertyConstantValue(ConstantValue, CaseProperty):
     pass
 
 
+@attr.s(auto_attribs=True, kw_only=True)
+class SupercaseValueSource(ValueSource):
+    """
+    A reference to a list of parent/host cases.
+
+    Evaluates nested ValueSource config, allowing for recursion.
+    """
+    supercase_value_source: dict
+
+    # Optional filters for indices
+    identifier: Optional[str] = None
+    referenced_type: Optional[str] = None
+    # relationship: Optional[Literal['child', 'extension']] = None  # Py3.8+
+    relationship: Optional[str] = None
+
+    @classmethod
+    def get_schema_params(cls) -> Tuple[Tuple, Dict]:
+        (schema, *other_args), kwargs = super().get_schema_params()
+        schema.update({
+            'supercase_value_source': And(dict, len),
+            SchemaOptional('identifier'): str,
+            SchemaOptional('referenced_type'): str,
+            SchemaOptional('relationship'): lambda r: r in ('child', 'extension'),
+        })
+        return (schema, *other_args), kwargs
+
+    def get_commcare_value(self, info):
+        values = []
+        for supercase_info in self._iter_supercase_info(info):
+            value_source = as_value_source(self.supercase_value_source)
+            value = value_source.get_commcare_value(supercase_info)
+            values.append(value)
+        return values
+
+    def get_import_value(self, external_data):
+        # OpenMRS Atom feed and FHIR API must build case blocks for
+        # related cases for this to be implemented
+        raise NotImplementedError
+
+    def set_external_value(self, external_data, info):
+        for i, supercase_info in enumerate(self._iter_supercase_info(info)):
+            value_source = as_value_source(self.supercase_value_source)
+            _subs_counters(value_source, i)
+            value_source.set_external_value(external_data, supercase_info)
+
+    def _iter_supercase_info(self, info: CaseTriggerInfo):
+
+        def filter_index(idx):
+            return (
+                (not self.identifier or idx.identifier == self.identifier)
+                and (not self.referenced_type or idx.referenced_type == self.referenced_type)
+                and (not self.relationship or idx.relationship == self.relationship)
+            )
+
+        case_accessor = CaseAccessors(info.domain)
+        case = case_accessor.get_case(info.case_id)
```

Comment:
This looks like it will always fetch the case from the db. Is there a way to do query caching if a set of multiple value sources operates on the same case rather than fetching a fresh copy for each one? Does that question even make sense?

Reply:
It does make sense. `quickcache()` with domain and case ID as keys?

Reply:
Hmmm. I'm a little worried about the risk of a stale cache. If a case is a Patient resource, and we register that Patient, then we will save its FHIR ID to `external_id`. We might need a fresh version if that case is the parent of a different resource. What do you think?

Reply:
I think I will need more context to say anything meaningful, so you probably have better instincts for how that should be designed than I do. I'd be happy to go to voice and talk it over if that would be helpful